### PR TITLE
Lower initial data target

### PR DESCRIPTION
**Pull request by jeanPerier**
Create a function `genInitialDataTarget` to create a `fir.box` for a pointer initial data target. In `defineGlobal` and `defineCommon`, use it if the global (or common member) is a pointer with an initial data target.

This is not added to `defineGlobalAggregateStore` because `C8106` prevents pointer/derived type with pointer component to be used in equivalences.
The fact that the initial data target cannot be in an equivalence (C8108 + C765) also simplifies the logic (there is no need to try instantiating aggregate stores inside global op initializer).
The main Fortran difficulty was that the initial data target can be in common block, and so can the pointer. This is covered by the `instantiateVar(var)` call on the initial data target that will declare the common block if needed (as this call it does in normal context).

Most of the complexity of  `genInitialDataTarget` stems from the fact that it is not possible to use `fir.rebox` in a global initializer (which we normally used in some pointer assignment case). I tried making this possible, but it was driving too much weird changes in `fir.rebox` codegen to support both an in memory and a struct descriptor argument. So I went for an ad-hoc handling of this corner case.

The last test cases `block data snake` and `block data tied` in `pointer-initial-target-2.f90` lowers fine to FIR and LLVM dialect, but crashes the LLVM dialect to LLVM IR pass. I am working on a simple fix for this, the issue turned out to be that the translation passes declare global op and define their body in the same pass, which is an issue if they are forward reference from an initializer to a globalOp not yet declared in LLVM IR.

Derived type with length parameters are left TODOs. I can't prepare much here without more general work on them.

### CODE REVIEW

File: flang/lib/Lower/ConvertVariable.cpp [hunk 1]
```diff
@@ -123,6 +123,105 @@ static fir::GlobalOp declareGlobal(Fortran::lower::AbstractConverter &converter,
...
+      return Fortran::semantics::CountLenParameters(*derived) > 0;
+  return false;
+}
+
+static mlir::Type unwrapElementType(mlir::Type type) {
+  if (auto ty = fir::dyn_cast_ptrOrBoxEleTy(type))
+    type = ty;
+  if (auto seqType = type.dyn_cast<fir::SequenceType>())
+    type = seqType.getEleTy();
+  return type;
```
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertVariable.cpp:140]: I'm surprised this isn't already implemented someplace. Seems to pop up often.

**Review by schweitzpgi — APPROVED** [review]
LGTM.

File: flang/lib/Lower/ConvertVariable.cpp (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Lower/ConvertVariable.cpp:140]: Right, we had a discussion a while ago about having this in `FIRType.h`. For types that wraps character/record, we might want quick access on the character/record type to take decisions based on the length parameters. The conclusion at the time was that it was too custom (so there is something very similar in CharacterExpr.cpp).
Would you be in favor of moving this in `FIRType.h` ?

Comment (schweitzpgi) [reply flang/lib/Lower/ConvertVariable.cpp:140]: Seems like the best place for it, if it can reduce some replication.
